facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+use crate::{
+    config::GCSPath,
+    transport::{Transport, TransportWriter},
+    Error,
+};
+use anyhow::{anyhow, Context, Result};
+use chrono::{prelude::Utc, DateTime, Duration};
+use serde::Deserialize;
+use std::{
+    io::{Cursor, Read, Write},
+    mem,
+};
+
+const STORAGE_API_BASE_URL: &str = "https://storage.googleapis.com";
+const DEFAULT_OAUTH_TOKEN_URL: &str =
+    "http://metadata.google.internal:80/computeMetadata/v1/instance/service-accounts/default/token";
+
+// A wrapper around an Oauth token and its expiration date.
+#[derive(Debug)]
+struct OauthToken {
+    token: String,
+    expiration: DateTime<Utc>,
+}
+
+impl OauthToken {
+    // Returns true if the token is not yet expired.
+    fn expired(&self) -> bool {
+        Utc::now() < self.expiration
+    }
+}
+
+// Represents the response from a GET request to the GKE metadata service's
+// service account token endpoint. Structure is derived from empirical
+// observation of the JSON scraped from inside a GKE job.
+#[derive(Debug, Deserialize, PartialEq)]
+struct MetadataServiceTokenResponse {
+    access_token: String,
+    expires_in: i64,
+    token_type: String,
+}
+
+// Represents the response from a POST request to the GCP IAM service's
+// generateAccessToken endpoint.
+// https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+struct GenerateAccessTokenResponse {
+    access_token: String,
+    expire_time: DateTime<Utc>,
+}
+
+// OauthTokenProvider manages a default service account Oauth token (i.e. the
+// one for a GCP service account mapped to a Kubernetes service account) and an
+// Oauth token used to impersonate another service account.
+struct OauthTokenProvider {
+    service_account_to_impersonate: Option<String>,
+    default_service_account_oauth_token: Option<OauthToken>,
+    impersonated_service_account_oauth_token: Option<OauthToken>,
+}
+
+impl OauthTokenProvider {
+    // Creates a token provider which can impersonate the specified service
+    // account.
+    fn new(service_account_to_impersonate: Option<String>) -> OauthTokenProvider {
+        OauthTokenProvider {
+            service_account_to_impersonate,
+            default_service_account_oauth_token: None,
+            impersonated_service_account_oauth_token: None,
+        }
+    }
+
+    fn ensure_storage_access_oauth_token(&mut self) -> Result<String> {
+        match self.service_account_to_impersonate {
+            Some(_) => self.ensure_impersonated_service_account_oauth_token(),
+            None => self.ensure_default_service_account_oauth_token(),
+        }
+    }
+
+    // Obtains an Oauth token for the default service account, if one is not
+    // already held or if it has expired. Otherwise returns a copy of the token.
```

Comment:
The wording here is pretty confusing - what token is it referring to in the "otherwise" clause?

How about:

Returns the current OAuth token for the default service account, if it is valid. Otherwise obtains and returns a new one.